### PR TITLE
Add meson build system

### PR DESCRIPTION
**Pull request by Faless**
In this PR:

- The work in #45093 .
- The work in #51153 .
- Restored all the modified godot files.
- Add a wrapper script to run scons builders (much less code duplication)
- Add CI to test on various platforms.

Missing modules:
`raycast, webm, opus, webm, theora, gdnative, etcpak, freetype, mono`.

`text_server_adv` fails to link.

Additionally, at least the X11 display driver does not work (fails to connect to display).
The windows build is broken because... well, I don't know, meson doesn't seem to get the path slashes right.

This is as far as I will go, I already spent 3 days fighting with this  and I'm totally fed up.
If anyone want to keep up the work, I think this is a solid base which is also easy to rebase.

Note: some cleanup of the doc generation should fix the mac editor build. I'll try to extract some of the scons code and move it to editor_generators.py as it should be.
Note2: I don't think `iphone` is actually a valid iphone build.

### CODE REVIEW

**Comment by fire** [comment]
My comment is if we can run meson side by side with scons, we can work on this incrementally.

Need to develop a plan with Akien though.

**Comment by Faless** [comment]
@fire yes, that was my idea, and the reason with I reworked the previous PR bringing the sources back in sync.

If anyone can help with the windows and osx builds it would be great.
I can also simplify the CI and have a single "meson" workflow that runs all the platforms.

File: core/meson.build [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+version_cdata = configuration_data()
+version_cdata.set_quoted('VERSION_SHORT_NAME', 'Godot')
+version_cdata.set_quoted('VERSION_NAME', 'Godot Engine')
+version_cdata.set('VERSION_MAJOR', version_arr[0])
+version_cdata.set('VERSION_MINOR', version_arr[1])
+version_cdata.set('VERSION_PATCH', version_arr[2])
+version_cdata.set_quoted('VERSION_STATUS', 'dev')
+version_cdata.set_quoted('VERSION_BUILD', 'custom_build')
+version_cdata.set_quoted('VERSION_MODULE_CONFIG', '')
+version_cdata.set('VERSION_YEAR', 2020)
```
Comment (AnilBK) [inline core/meson.build:55]: *2021.

File: version.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "short_name": "godot",
+    "name": "Godot Engine",
+    "major": 4,
+    "minor": 0,
+    "patch": 0,
+    "status": "dev",
+    "module_config": "",
+    "year": 2020,
```
Comment (AnilBK) [inline version.json:9]: Same here. 2021.

Comment (AnilBK) [inline version.json:9]: Same here.
2021.

File: .github/actions/meson_install/action.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+name: 'Setup python meson ninja.'
```
Comment (aaronfranke) [inline .github/actions/meson_install/action.yml:1]: Should these be capitalized? Also, the quotes are inconsistent. Other places in the file use `'`, `"`, or no quotes.

**Comment by bruvzg** [comment]
> `text_server_adv` fails to link.

It's linking for me (on macOS), but generated `register_module_types.gen.cpp` is empty, and it's not registered properly (I guess all other modules are broken for the same reason).

**Comment by Faless** [comment]
@bruvzg nicely spot! Should be fixed now. Let's see what happens to CI

**Comment by bruvzg** [comment]
Another issue, disabling module doesn't disable its dependencies (at least true for `denoise` → `oidn` which is not supported on ARM64 and breaks M1 mac build).

**Comment by Listwon** [comment]
> The windows build is broken because... well, I don't know, meson doesn't seem to get the path slashes right.

@Faless I spent yesterday evening trying to figure out what's wrong and surprisingly it's not the problem with slashes but the Python shebang missing in affected Python scripts. I moved the files around and for unknown reason `version.py` worked while `glsl_builders.py` and some others didn't. Adding `#!/usr/bin/python3` in `glsl_builders.py` (the same as in `version.py`) fixed `find_program()` issue and meson setup moved to the next script in order - `default_theme_builders.py`, also missing shebang.

**Comment by Faless** [comment]
@Listwon thanks a lot for finding that out. I'm growing more and more disappointed with meson :/ .
But let's see how this goes.

**Comment by Faless** [comment]
Added volk support, now the editor starts (on X11 at least).

Font rendering is totally broken, I realize that FreeType does actually get built (not the module though, because the module in scons only existed to let the library build).

So I think it might be some missing defines that causes the font rendering to be broken.

**Comment by reduz** [comment]
I suggest the path forward with Meson would be as follows:
* Merge this into a "meson" branch in godotengine/godot
* Try to get it into a more stable state, something we could call "beta". This means all platforms working and all modules working, minor issues maybe remaining.
* Merge the branch into master and ask contributors to test it while the remaining minor issues are resolved. At this stage, any change to the build system via PR should be made for SCons and Meson, so the idea is that this time of co-existence is as minimal as possible.
* Deprecate SCons

Does this make sense?

**Comment by jpakkane** [comment]
With this diff I can build this branch on Linux and run the result successfully:

```diff
diff --git a/modules/text_server_adv/text_server_adv.h b/modules/text_server_adv/text_server_adv.h
index 5989035800..a9a29fd35c 100644
--- a/modules/text_server_adv/text_server_adv.h
+++ b/modules/text_server_adv/text_server_adv.h
@@ -52,6 +52,7 @@
 #include <unicode/uloc.h>
 #include <unicode/uscript.h>
 #include <unicode/ustring.h>
+#include <unicode/utf16.h>
 #include <unicode/utypes.h>
 
 #include "modules/modules_enabled.gen.h"
diff --git a/thirdparty/harfbuzz/meson.build b/thirdparty/harfbuzz/meson.build
index caca75438e..e9b282854d 100644
--- a/thirdparty/harfbuzz/meson.build
+++ b/thirdparty/harfbuzz/meson.build
@@ -73,7 +73,7 @@ if get_option('builtin_harfbuzz')
         '-DHAVE_ICU',
         '-DHAVE_FREETYPE',
         '-DHAVE_GRAPHITE2',
-        '-DHB_DISABLE_DEPRECATED',
+        #'-DHB_DISABLE_DEPRECATED',
         '-DGRAPHITE2_STATIC',
     ]
```

Further, if you add [this commit](https://github.com/jpakkane/godot/commit/7997e23111932cad8bf2480f48edc62d90680d60) the ordering issues on macOS should get fixed.

**Comment by jpakkane** [comment]
With this change it builds and runs on Windows, too:

```
diff --git a/modules/camera/meson.build b/modules/camera/meson.build
index b5f4457b67..26c15871ed 100644
diff --git a/thirdparty/etcpak/meson.build b/thirdparty/etcpak/meson.build
index 6fc6f41d65..ccb6fb959c 100644
--- a/thirdparty/etcpak/meson.build
+++ b/thirdparty/etcpak/meson.build
@@ -6,5 +6,8 @@ _etcpak_srcs = files([
     'Tables.cpp',
 ])

-_lib_etcpak = static_library('etcpak', _etcpak_srcs, build_by_default: false)
+_lib_etcpak = static_library('etcpak',
+                             _etcpak_srcs,
+                             cpp_args: '-DNOMINMAX',
+                             build_by_default: false)
 DEP_ETCPAK = declare_dependency(link_with: _lib_etcpak)

diff --git a/thirdparty/oidn/common/platform.h b/thirdparty/oidn/common/platform.h
index 9373b617b5..a08cdf39e4 100644
--- a/thirdparty/oidn/common/platform.h
+++ b/thirdparty/oidn/common/platform.h
@@ -18,7 +18,9 @@

 #if defined(_WIN32)
   #define WIN32_LEAN_AND_MEAN
-  #define NOMINMAX
+  #ifndef NOMINMAX
+    #define NOMINMAX
+  #endif
   #include <windows.h>
 #elif defined(__APPLE__)
   #include <sys/sysctl.h>
```

**Comment by Faless** [comment]
@jpakkane the whole point of this branch is not changing the C/C++ source.
If it builds with scons, it should build with meson without having to change the source files. Especially third party libraries code.

I'll try to pick the `meson.build` commits for now.

**Comment by jpakkane** [comment]
With the following patch this builds on Windows for me:

```diff
diff --git a/thirdparty/etcpak/meson.build b/thirdparty/etcpak/meson.build
index 6fc6f41d65..7c6cd63c35 100644
--- a/thirdparty/etcpak/meson.build
+++ b/thirdparty/etcpak/meson.build
@@ -1,10 +1,12 @@
 # the include directory
-_etcpak_srcs = files([
+_etcpak_srcs = files(
     'Dither.cpp',
     'ProcessDxtc.cpp',
     'ProcessRGB.cpp',
     'Tables.cpp',
-])
+)

-_lib_etcpak = static_library('etcpak', _etcpak_srcs, build_by_default: false)
+_lib_etcpak = static_library('etcpak',
+    _etcpak_srcs,
+    cpp_args: ['-DNOMINMAX'])
 DEP_ETCPAK = declare_dependency(link_with: _lib_etcpak)
```

**Comment by jpakkane** [comment]
This patch is needed to build with Visual Studio. It fails with the same Freetype error as the other ones.

```diff
diff --git a/tests/meson.build b/tests/meson.build
index 8bb59768c6..6f4481c7a1 100644
--- a/tests/meson.build
+++ b/tests/meson.build
@@ -3,7 +3,11 @@ if get_option('tests')

     _tests_cpp_args = []
     if PLATFORM == 'windows'
+      if cpp.get_id() == 'msvc'
+        _tests_cpp_args += '/bigobj'
+      else
         _tests_cpp_args += '-DDOCTEST_THREAD_LOCAL'
+      endf
     endif

     #_tests_incdirs = include_directories('.')
```

**Comment by jpakkane** [comment]
In order to set up the VS environment on Github Actions, you need to use [this action](https://github.com/ilammy/msvc-dev-cmd). From what I can tell, adding one line in `.github/workflows/meson.yml` line 106 should do it:

```
  - uses: actions/checkout@v2
  - uses: ilammy/msvc-dev-cmd@v1
```

Scons seems to autoactivate the VS environment. Meson also has code for that, but it only does it if it detects that no other build toolchain is currently available. The assumption here being that if you have added a build toolchain in your 'PATH` then that is probably what you want to use.

**Comment by lightspot21** [comment]
Any progress on this?

**Comment by underdoeg** [comment]
Is this going somwhere? The release of 4.0 would be a great moment to introduce another build system.

**Comment by akien-mga** [comment]
I'll share what I wrote on the Godot chat for context:

> **jpakkane |** There was some talk that the Meson branch would receive proper review after the alfa is out. What's the status for that (the branch is up to date as of a few days ago).
>
> **Akien |** @jpakkane That was my expectation but I see now that it will take longer. I really need to be able to dedicate several days to reviewing your work, learning Meson at the same time, and assess what's the status for the many build configurations that we currently support (e.g. building for macOS and iOS from Linux with OSXCross, MinGW support with GCC and Clang from Windows, Linux and macOS, Mono builds, etc.).
That's a ton of work that I still can't allocate right now while all my time is taken by maintaining three branches in parallel (master, 3.x, 3.4).
Godot 4.0 is a very complex release and we really need to narrow down on bugfixing now so that we can get the release out. I hope to find time to review and assess Meson as a potential replacement for our current buildsystem before 4.0 is released, but at this stage I can't make promises. Changing the buildsystem is not something that we **need** to do right now, and I can't allocate much time to "nice to have"s at this time (especially right now where reduz has been AWOL for a couple months due to moving countries so I've been leading development mostly alone).
Of course there's always the option discussed a while ago to merge the Meson port in a dedicated feature branch, and let interested contributors work on to bring it to complete feature parity with SCons. But I don't feel right encouraging this work and letting many contributors invest time in it if we haven't decided loud and clear to change the buildsystem - and to decide this I need to spend time with it and confirm that it can do everything we currently do, do it better, and do it with a nicer implementation.
>
> **jpakkane |** Is there anything I can do to make it easier?
>
> **Akien |** @jpakkane Probably not right now, though your periodical rebases are appreciated! I don't want to give you more work before I'm able to do my part - especially before I can confirm that we want to proceed with this and commit to merging it as a replacement for the current buildsystem. Once I _do_ spend the needed time to get comfortable with Meson and review the current state, I'll surely have a number of things I'll ask of you and other interested contributors to bring it to full feature parity. (And do my fair share to also get familiar with the implementation so that I'm able to maintain it.)


Note that @jpakkane has been maintaining a branch against current `master` here for those interested in testing / giving feedback: https://github.com/jpakkane/godot/tree/mesoncitest2

**Comment by Faless** [comment]
Closing, as development is continuing in #51153

**Pull request closed without merging**